### PR TITLE
Introduce `HistoricalBuild` interface

### PR DESCRIPTION
**Pull request by Vlatombe**
This allows to display `Run`-like objects in build history without having concrete `Run` present in the Jenkins instance.

* Related soft downstream https://github.com/jenkinsci/script-security-plugin/pull/576

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Run.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Run.java:702]: Moved to `HistoricalBuild`

File: core/src/main/java/hudson/model/Run.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Run.java:2170]: Moved to `HistoricalBuild`

File: core/src/main/java/hudson/model/Run.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Run.java:171]: Inlined in `HistoricalBuild#getTruncatedDescription`

File: core/src/main/resources/hudson/widgets/BuildHistoryWidget/entries.jelly
Comment (Vlatombe) [inline core/src/main/resources/hudson/widgets/BuildHistoryWidget/entries.jelly:1]: This file was actually dead code since https://github.com/cloudbees/private-jenkins/commit/078d3c7c4b5ffc5eae2f8c300a681bc340dfec60

File: core/src/main/java/jenkins/widgets/HistoryPageFilter.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/HistoryPageFilter.java:288]:
```suggestion
        } else if (entry instanceof HistoricalBuild run) {
```

File: core/src/main/java/jenkins/widgets/HistoryPageFilter.java
Comment (jglick) [inline core/src/main/java/jenkins/widgets/HistoryPageFilter.java:246]:
```suggestion
        if (widget != null && widget.owner instanceof Job job) {
            return job.getNextBuildNumber();
```

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/model/Run.java
Comment (Vlatombe) [reply core/src/main/java/hudson/model/Run.java:502]: Since `getActions()` doesn't exist on `HistoricalBuild`, I left it there.

**Review by alecharp — APPROVED** [review]

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by jglick — APPROVED** [review]

**Pull request merged**
